rfcs/20180905-deprecate-collections.md

```diff
@@ -0,0 +1,575 @@
+
+
+### **Moving average variables**
+
+Collections are used in two places here. We'll remove these calls and ask callers to explicitly pass in a variable list 
+
+
+
+*   [MovingAverageOptimizer.swapping_saver](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/contrib/opt/python/training/moving_average_optimizer.py#L140)
+*   [ExponentialMovingAverage.variables_to_restore](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/training/moving_averages.py#L473). We'll have to expose the variables created by ExponentialMovingAverage as they would have to be passed in here for this method.
+
+
+### **Tables**
+
+
+#### **Keras models**
+
+We'll provide a `track_table` method in the backend (similar to `track_variable`) and we'll make sure that when model.fit is called, tables are initialized similar to the way we initialize variables. Changes we'll have to make here are the following
+
+
+
+*   Add a `track_table` method similar to `track_variable` method in keras backend.py
+*   Change feature column code to call `track_table` when it creates tables
+*   Modify `get_session` to initialize the tracked tables in addition to variables
+
+
+#### **Custom graphs**
+
+Similar to `variable_creator_scope` we'll create a `table_creator_scope` that will allow tracking logic to be introduced each time a table is created. This can then be used to collect up all the tables. 
+
+Sample user code might look like
+
+
+```python
+table_creator = tf.TableCreator()
+
+with table_creator:
+  … <graph creation code>
+  table = tf.HashTable(tf.KeyValueTensorInitializer(keys, values), -1)
+  … 
+
+assert table_creator.tables() == [table]
+```
+
+
+We can add some helper methods on TableCreator to collect up all `initializers, vocabulary_filenames`.
+
+
+#### **Alternative proposal for tables**
+
+Initialize tables on first use. Similar to the [OneShotIterator](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/core/ops/dataset_ops.cc#L684), we'll pass in a function argument that can initialize the table. We can even explore the idea of the function to return a dataset and then initialize the table from that dataset. Having a function would avoid overheads of passing in large tensors that might be required to initialize on each lookup (index_table_from_tensor for instance might run into this problem). The Table API would remain the same since we hide initialization from the user and we wouldn't have to run an initializer explicitly. 
+
+The concern with this solution is that in serving, the first request would take more time than the rest of the calls and that sort of inconsistent performance is not desirable. 
+
+
+### **Asset file paths**
+
+Asset files currently only include vocabulary files that are used to initialize tables with. The tracking of tables should be sufficient to track vocabulary files. 
+
```

Comment:
These also have to be replicated in the SavedModel, although passing these in as a python list would be way clearer than the strange thing we do now with collections, presumably. But it will require updating client code and SavedModel code as well.